### PR TITLE
Render counter widgets using relative font size

### PR DESCRIPTION
**Pull request by eradman**
## What type of PR is this? 

- [x] Adjustment to the UI

## Description

Using `transform` scales contents in in height and width, resulting in text wrapping even if there is horizontal space.

## How is this tested?

- [x] Manually

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)

**Screen shot for master**

counter_display_orig

**Screen shot for this PR**

counter_display_new

### CODE REVIEW

**Comment by eradman** [comment]
Two frontend checks failed

* "sets the correct height of table visualization:"
* "Auto height for table visualization"

Should be easy to adjust the assertions in `widget_spec.js` if the overall goal of this PR is approved

**Comment by guidopetri** [comment]
I agree the second screenshot you posted looks better. What do the (failing) tests show before/after the change?

I don't have the necessary context to review the PR - I'm not good at JS :) @getredash/maintainers anyone comfortable with these changes?

**Comment by justinclift** [comment]
@gabrieldutra Do you have time to take a look at this? :smile:

**Comment by justinclift** [comment]
Hmmm, it's failed the CI tests twice in the same spot.  That *probably* means the tests need updating to match this PR.

```
   7 passing (1m)
  2 failing

  1) Widget
       sets the correct height of table visualization:

      Timed out retrying
      + expected - actual

      -387
      +[ 380, 381 ]
      
      at Context.eval (http://server:5000/__cypress/tests?p=client/cypress/integration/dashboard/widget_spec.js:242:66)

  2) Widget
       Auto height for table visualization
         Height behavior on refresh
           grows when dynamically adding table rows:

      Timed out retrying
      + expected - actual

      -385
      +435
      
      at Context.eval (http://server:5000/__cypress/tests?p=client/cypress/integration/dashboard/widget_spec.js:203:44)
```

Though it could also mean we just got unlucky in the two test runs.

I'll start the testing again, though I kind of expect it's probably showing a real "issue", so the tests will probably need updating. :smile:

**Comment by justinclift** [comment]
Nope, it just failed again in the same place.

@eradman Are you ok to look into those test failures?  If you're not already familiar with Cypress then there are are initial instructions for running the tests locally, here:

https://github.com/getredash/redash/wiki/Testing-your-changes#comprehensive-e2e-testing-cypress

**Comment by justinclift** [comment]
@eradman When that happens to me (tests fail in CI but not locally), it seems to mean I've been testing with a cached (working) docker image.

I'd probably try something like:

## 1. Make sure all docker containers have been stopped

```
$ make down   <-- stops containers started by "make test"
$ yarn cypress stop    <-- stops containers started by "yarn cypress run"
```

Confirm there aren't any docker containers still running:

```
$ docker ps
CONTAINER ID   IMAGE     COMMAND   CREATED   STATUS    PORTS     NAMES
$
```

## 2. Force remove all the cypress_* and redash_* docker images

First get the list of docker images.  For example, on my system right now:

```
$ docker image ls
REPOSITORY                    TAG               IMAGE ID       CREATED          SIZE
pgautoupgrade/pgautoupgrade   dev               7680230c3094   50 minutes ago   377MB
<none>                        <none>            c3c0b815f985   50 minutes ago   1.74GB
cypress_scheduler             latest            fc535303c776   5 hours ago      1.06GB
cypress_server                latest            fc535303c776   5 hours ago      1.06GB
cypress_worker                latest            fc535303c776   5 hours ago      1.06GB
<none>                        <none>            e78bcb9489b0   5 hours ago      929MB
redash_scheduler              latest            086c3a87816c   5 hours ago      1.15GB
redash_server                 latest            086c3a87816c   5 hours ago      1.15GB
redash_worker                 latest            086c3a87816c   5 hours ago      1.15GB
```

Then force remove the ones whose repo names start with `cypress_` or `redash_`:

```
$ docker image rm --force fc535303c776 086c3a87816c
```

## 3. Then rebuild your Cypress test image and start the tests again

Then follow the steps for building and running Cypress tests:

```
$ yarn cypress build  <-- creates the cypress docker image
$ yarn cypress start  <-- starts the cypress containers using that image
$ yarn cypress run    <-- runs the cypress tests
```

**Comment by eradman** [comment]
@justinclift indeed removing existing docker containers was the reason tests were passing for me locally.

Rebased this PR to pick up the update to Cypress to 6.0.1

File: viz-lib/src/visualizations/counter/Renderer.tsx [hunk 1]
```diff
@@ -10,17 +10,19 @@ import "./render.less";
...
   };
 }
 
 function getCounterScale(container: any) {
-  const inner = container.firstChild;
-  const scale = Math.min(container.offsetWidth / inner.offsetWidth, container.offsetHeight / inner.offsetHeight);
-  return Number(isFinite(scale) ? scale : 1).toFixed(2); // keep only two decimal places
+  // size of font in base container
+  // children use a relative font size (em)
+  if (container.clientHeight < 50)
```
Comment (gaecoli) [inline viz-lib/src/visualizations/counter/Renderer.tsx:20]: Maybe fixed value is not a better way.

Comment (justinclift) [reply viz-lib/src/visualizations/counter/Renderer.tsx:20]: Oh, that's an interesting idea.  That would probably be more flexible.

We can try it I guess? :smile:

Comment (eradman) [reply viz-lib/src/visualizations/counter/Renderer.tsx]: @gaecoli I adjusted the logic to scale the text using the container height, but with fixed upper and lower boundaries. Let me know what you think of this.

**Comment by eradman** [comment]
I dropped this part of the PR since it is causing the tests to fail. It probably belongs in a separate change because it affects more than just counter widgets:

```diff
--- a/client/app/components/dashboards/dashboard-widget/Widget.less
+++ b/client/app/components/dashboards/dashboard-widget/Widget.less
@@ -218,7 +218,7 @@
   }

   .t-header.widget {
-    padding: 15px;
+    padding: 12px;
   }

   &:hover,
@@ -237,7 +237,7 @@
   }

   .tile__bottom-control {
-    padding: 10px 15px;
+    padding: 10px 12px;
     display: flex;
     justify-content: space-between;
     align-items: center;
```

**Comment by eradman** [comment]
Updated screenshots at the top of this PR

**Comment by eradman** [comment]
Adding another screenshot to show the practical effect of scaling font-size by height while allowing the text to wrap horizontally
counter-vis-comparison

_I did find one oddity:_ when editing the visualization the `counter-visualization-container` is only 68px so the text is rendered at the smallest size

**Comment by justinclift** [comment]
Visually, those look fine to me personally.

I'm not understanding what that oddity is about though.  Does that mean more investigation needs to be done?

**Comment by eradman** [comment]
Yes, I need to fix the scalling in the visualization editor. It's usable, but I expect the preview to expand. Screenshot:

counter-vis-editor

**Comment by justinclift** [comment]
Oh.  Yeah, it'd be best to fix that. :smile:

**Comment by eradman** [comment]
Latest revision addresses the issue with the visualization preview

**Comment by justinclift** [comment]
Awesome!  Good stuff @eradman. :smile:

**Comment by myonlylonely** [comment]
It seems this commit let the counter number became extremely small.https://github.com/getredash/redash/issues/6521

**Pull request merged**